braindecode/preprocessing/eegprep_preprocess.py

```diff
@@ -0,0 +1,1277 @@
+        Whether to record the EEG channel locations before processing
+        in the MNE Raw structure for later use. This will not override any already
+        present channel location information, so this can safely be used multiple times
+        to record whatever were the original channel locations.
+    force_dtype : np.dtype | str | None
+        Optionally for the in/out data to be converted to this dtype before and after
+        processing. Can help ensure consistent dtypes across different preprocessors.
+
+    """
+
+    # payload key under which we store our original channel locations
+    _chanlocs_key = "original_chanlocs"
+
+    def __init__(
+        self,
+        *,
+        limit_threads: int | None = 4,
+        can_change_duration: str | bool | None = None,
+        record_orig_chanlocs: bool = False,
+        force_dtype: np.dtype | str | None = None,
+    ):
+        super().__init__(
+            fn=self._apply_op,
+            apply_on_array=False,
+        )
+        if can_change_duration is True:
+            can_change_duration = self.__class__.__name__
+        elif can_change_duration is False:
+            can_change_duration = None
+        self.can_change_duration = can_change_duration
+        self.limit_threads = limit_threads
+        self.record_orig_chanlocs = record_orig_chanlocs
+        self.force_dtype = np.dtype(force_dtype) if force_dtype is not None else None
+        self._raw: BaseRaw | None = None
+
+    def apply_eeg(self, EEG: dict[str, Any]) -> dict[str, Any]:
+        """Apply the preprocessor to an EEGLAB EEG structure. Overridden by subclass."""
+        raise NotImplementedError("Subclasses must implement apply_eeg().")
+
+    def _apply_op(self, raw: BaseRaw) -> None:
+        """Internal method that does the actual work; this is called by Preprocessor.apply()."""
+        opname = self.__class__.__name__
+        if isinstance(raw, BaseEpochs):
+            raise ValueError(
+                f"{opname} is meant to be used on Raw (continuous) data, "
+                f"not Epochs. Use before epoching."
+            )
+
+        # split off non-EEG channels since the op doesn't handle those
+        EEG, nonEEG, chn_order = _separate_noneeg_channels(raw)
+
+        # convert to EEGLAB structure and prepare for processing
+        EEG = _mne2eegprep(EEG)
+
+        orig_chanlocs = list(EEG["chanlocs"])
+
+        with _maybe_threadpool_limits(nthreads=self.limit_threads):
+            try:
+                self._raw = raw
+
+                if self.force_dtype is not None:
+                    EEG["data"] = EEG["data"].astype(self.force_dtype)
+
+                # actual operation happens here
+                EEG = self.apply_eeg(EEG)
+
+                if self.force_dtype is not None:
+                    EEG["data"] = EEG["data"].astype(self.force_dtype)
+
+            finally:
+                # clear again so we don't keep a reference around
+                self._raw = None
```

Comment:
I think this is will be a little complicated because we have to use fork, process, and thread, but we can discuss in the second iteration of the PR

Reply:
Removed the use of the `self._raw` mutable state, assuming that's the concern.